File: LeetCode/Trees/102/BinaryTreeLevelOrderTraversal.cs
// LeetCode #102
// https://leetcode.com/problems/binary-tree-level-order-traversal/
//
// Given a binary tree, return the level order traversal of its nodes' values.

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// Level Order Traversal of a tree is essentially a Breadth First Search (BFS).
// The standard BFS implementation uses a queue, where as every node is visited all of its adjacent nodes get added to the queue for future processing.

 // Time: O(n)
 // Space: O(w) where w is the max width of the tree
public class Solution {
    public IList<IList<int>> LevelOrder(TreeNode root) {
        List<IList<int>> valueList = new List<IList<int>>();
        if (root == null) return valueList;

        Queue<TreeNode> queue = new Queue<TreeNode>();
        queue.Enqueue(root);

        while (queue.Count != 0) {
            int count = queue.Count;
            List<int> levelValueList = new List<int>();
            for (int i = 0; i < count; i++) {
                TreeNode currentNode = queue.Dequeue();
                levelValueList.Add(currentNode.val);
                if (currentNode.left != null) {
                    queue.Enqueue(currentNode.left);
                }
                if (currentNode.right != null) {
                    queue.Enqueue(currentNode.right);
                }
            }
            valueList.Add(levelValueList);
        }

        return valueList;
    }
}

File: Solutions/GrokkingTheCodingInterview/TreeBreadthFirstSearch/BinaryTreeLevelOrderTraversal.java
// From the educative.io course "Grokking the Coding Interview".
// Pattern: Tree Breadth First Search
// Problem: Binary Tree Level Order Traversal (easy)
// Related LeetCode Problem: https://leetcode.com/problems/binary-tree-level-order-traversal/

// Problem Statement:
// Given a binary tree, populate an array to represent its level-by-level traversal. 
// You should populate the values of all nodes of each level from left to right in separate sub-arrays.

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

// Time: O(n) where n is the number of nodes in the tree.
// Space: O(n) where n is the number of nodes in the tree.
public static List<List<Integer>> traverse(TreeNode root) {
    List<List<Integer>> result = new ArrayList<List<Integer>>();
    
    if (root == null) {
        return result;
    }

    Queue<TreeNode> nodeQueue = new LinkedList<TreeNode>();
    nodeQueue.offer(root);

    while (!nodeQueue.isEmpty()) {
        int levelNodeCount = nodeQueue.size();
        List<Integer> levelList = new ArrayList<Integer>();
        
        // Process all the nodes in this level.
        for (int i = 0; i < levelNodeCount; i++) {
            TreeNode currentNode = nodeQueue.poll();
            levelList.add(currentNode.val);
            
            if (currentNode.left != null) {
                nodeQueue.offer(currentNode.left);
            }
            if (currentNode.right != null) {
                nodeQueue.offer(currentNode.right);
            }
        }
        
        result.add(levelList);
    }

    return result;
}
